File: tests/node/domain/stack-mapping.test.ts
import { describe, it, expect } from "vitest";
import {
  LANGUAGE_COMMANDS,
  FRAMEWORK_PORTS,
  FRAMEWORK_HEALTH_PATHS,
  FRAMEWORK_LANGUAGE_RULES,
  NATIVE_SUPPORTED_FRAMEWORKS,
  VALID_INTERFACE_TYPES,
  VALID_ARCHITECTURE_STYLES,
  INTERFACE_SPEC_PROTOCOL_MAP,
  DOCKER_BASE_IMAGES,
  HOOK_TEMPLATE_MAP,
  SETTINGS_LANG_MAP,
  DATABASE_SETTINGS_MAP,
  CACHE_SETTINGS_MAP,
  DEFAULT_PORT_FALLBACK,
  DEFAULT_HEALTH_PATH,
  DEFAULT_DOCKER_IMAGE,
  getHookTemplateKey,
  getSettingsLangKey,
  getDatabaseSettingsKey,
  getCacheSettingsKey,
} from "../../../src/domain/stack-mapping.js";

describe("LANGUAGE_COMMANDS", () => {
  it("contains_8_entries", () => {
    expect(Object.keys(LANGUAGE_COMMANDS)).toHaveLength(8);
  });

  it.each([
    ["java-maven", "./mvnw compile -q", "./mvnw package -DskipTests", "./mvnw verify", "./mvnw verify jacoco:report", ".java", "pom.xml", "maven"],
    ["java-gradle", "./gradlew compileJava -q", "./gradlew build -x test", "./gradlew test", "./gradlew test jacocoTestReport", ".java", "build.gradle", "gradle"],
    ["kotlin-gradle", "./gradlew compileKotlin -q", "./gradlew build -x test", "./gradlew test", "./gradlew test jacocoTestReport", ".kt", "build.gradle.kts", "gradle"],
    ["typescript-npm", "npx --no-install tsc --noEmit", "npm run build", "npm test", "npm test -- --coverage", ".ts", "package.json", "npm"],
    ["python-pip", "python3 -m py_compile", "pip install -e .", "pytest", "pytest --cov", ".py", "pyproject.toml", "pip"],
    ["go-go", "go build ./...", "go build ./...", "go test ./...", "go test -coverprofile=coverage.out ./...", ".go", "go.mod", "go"],
    ["rust-cargo", "cargo check", "cargo build", "cargo test", "cargo tarpaulin", ".rs", "Cargo.toml", "cargo"],
    ["csharp-dotnet", "dotnet build --no-restore --verbosity quiet", "dotnet build", "dotnet test", 'dotnet test --collect:"XPlat Code Coverage"', ".cs", "*.csproj", "dotnet"],
  ])("%s_returnsCorrectCommands", (key, compileCmd, buildCmd, testCmd, coverageCmd, ext, buildFile, pkgMgr) => {
    const cmd = LANGUAGE_COMMANDS[key];
    expect(cmd).toBeDefined();
    expect(cmd!.compileCmd).toBe(compileCmd);
    expect(cmd!.buildCmd).toBe(buildCmd);
    expect(cmd!.testCmd).toBe(testCmd);
    expect(cmd!.coverageCmd).toBe(coverageCmd);
    expect(cmd!.fileExtension).toBe(ext);
    expect(cmd!.buildFile).toBe(buildFile);
    expect(cmd!.packageManager).toBe(pkgMgr);
  });
});

describe("FRAMEWORK_PORTS", () => {
  it("contains_11_entries", () => {
    expect(Object.keys(FRAMEWORK_PORTS)).toHaveLength(11);
  });

  it.each([
    ["quarkus", 8080],
    ["spring-boot", 8080],
    ["nestjs", 3000],
    ["fastapi", 8000],
    ["aspnet", 5000],
    ["axum", 3000],
  ])("%s_returnsPort_%i", (fw, port) => {
    expect(FRAMEWORK_PORTS[fw]).toBe(port);
  });

  it("defaultPortFallback_is8080", () => {
    expect(DEFAULT_PORT_FALLBACK).toBe(8080);
  });
});

describe("FRAMEWORK_HEALTH_PATHS", () => {
  it("contains_11_entries", () => {
    expect(Object.keys(FRAMEWORK_HEALTH_PATHS)).toHaveLength(11);
  });

  it("quarkus_hasSpecificPath", () => {
    expect(FRAMEWORK_HEALTH_PATHS["quarkus"]).toBe("/q/health");
  });

  it("springBoot_hasActuatorPath", () => {
    expect(FRAMEWORK_HEALTH_PATHS["spring-boot"]).toBe("/actuator/health");
  });

  it("defaultHealthPath_isSlashHealth", () => {
    expect(DEFAULT_HEALTH_PATH).toBe("/health");
  });
});

describe("FRAMEWORK_LANGUAGE_RULES", () => {
  it("contains_15_entries", () => {
    expect(Object.keys(FRAMEWORK_LANGUAGE_RULES)).toHaveLength(15);
  });

  it("quarkus_supportsJavaAndKotlin", () => {
    expect(FRAMEWORK_LANGUAGE_RULES["quarkus"]).toEqual(["java", "kotlin"]);
  });

  it("nestjs_supportsOnlyTypescript", () => {
    expect(FRAMEWORK_LANGUAGE_RULES["nestjs"]).toEqual(["typescript"]);
  });

  it("aspnet_supportsOnlyCsharp", () => {
    expect(FRAMEWORK_LANGUAGE_RULES["aspnet"]).toEqual(["csharp"]);
  });
});

describe("NATIVE_SUPPORTED_FRAMEWORKS", () => {
  it("containsQuarkusAndSpringBoot", () => {
    expect(NATIVE_SUPPORTED_FRAMEWORKS).toEqual(["quarkus", "spring-boot"]);
  });
});

describe("VALID_INTERFACE_TYPES", () => {
  it("contains_9_entries", () => {
    expect(VALID_INTERFACE_TYPES).toHaveLength(9);
  });

  it("includesExpectedTypes", () => {
    expect(VALID_INTERFACE_TYPES).toContain("rest");
    expect(VALID_INTERFACE_TYPES).toContain("grpc");
    expect(VALID_INTERFACE_TYPES).toContain("event-consumer");
    expect(VALID_INTERFACE_TYPES).toContain("scheduled");
  });
});

describe("VALID_ARCHITECTURE_STYLES", () => {
  it("contains_5_entries", () => {
    expect(VALID_ARCHITECTURE_STYLES).toHaveLength(5);
  });

  it("includesExpectedStyles", () => {
    expect(VALID_ARCHITECTURE_STYLES).toContain("microservice");
    expect(VALID_ARCHITECTURE_STYLES).toContain("library");
    expect(VALID_ARCHITECTURE_STYLES).toContain("serverless");
  });
});

describe("INTERFACE_SPEC_PROTOCOL_MAP", () => {
  it("mapsRestToOpenapi", () => {
    expect(INTERFACE_SPEC_PROTOCOL_MAP["rest"]).toBe("openapi");
  });

  it("mapsGrpcToProto3", () => {
    expect(INTERFACE_SPEC_PROTOCOL_MAP["grpc"]).toBe("proto3");
  });

  it("mapsEventConsumerToKafka", () => {
    expect(INTERFACE_SPEC_PROTOCOL_MAP["event-consumer"]).toBe("kafka");
  });
});

describe("DOCKER_BASE_IMAGES", () => {
  it("contains_7_entries", () => {
    expect(Object.keys(DOCKER_BASE_IMAGES)).toHaveLength(7);
  });

  it("javaUsesTemurin", () => {
    expect(DOCKER_BASE_IMAGES["java"]).toBe("eclipse-temurin:{version}-jre-alpine");
  });

  it("defaultDockerImage_isAlpine", () => {
    expect(DEFAULT_DOCKER_IMAGE).toBe("alpine:latest");
  });
});

describe("getHookTemplateKey", () => {
  it("javaMaven_returnsJavaMaven", () => {
    expect(getHookTemplateKey("java", "maven")).toBe("java-maven");
  });

  it("kotlinGradle_returnsKotlin", () => {
    expect(getHookTemplateKey("kotlin", "gradle")).toBe("kotlin");
  });

  it("pythonPip_returnsEmpty", () => {
    expect(getHookTemplateKey("python", "pip")).toBe("");
  });

  it("unknownCombo_returnsEmpty", () => {
    expect(getHookTemplateKey("unknown", "unknown")).toBe("");
  });
});

describe("getSettingsLangKey", () => {
  it("javaMaven_returnsJavaMaven", () => {
    expect(getSettingsLangKey("java", "maven")).toBe("java-maven");
  });

  it("kotlinGradle_returnsJavaGradle", () => {
    expect(getSettingsLangKey("kotlin", "gradle")).toBe("java-gradle");
  });

  it("unknownCombo_returnsEmpty", () => {
    expect(getSettingsLangKey("unknown", "unknown")).toBe("");
  });
});

describe("getDatabaseSettingsKey", () => {
  it("postgresql_returnsDatabasePsql", () => {
    expect(getDatabaseSettingsKey("postgresql")).toBe("database-psql");
  });

  it("mongodb_returnsDatabaseMongodb", () => {
    expect(getDatabaseSettingsKey("mongodb")).toBe("database-mongodb");
  });

  it("unknownDb_returnsEmpty", () => {
    expect(getDatabaseSettingsKey("unknown")).toBe("");
  });
});

describe("getCacheSettingsKey", () => {
  it("redis_returnsCacheRedis", () => {
    expect(getCacheSettingsKey("redis")).toBe("cache-redis");
  });

  it("unknownCache_returnsEmpty", () => {
    expect(getCacheSettingsKey("unknown")).toBe("");
  });
});

File: java/src/test/java/dev/iadev/domain/stack/StackMappingTest.java
package dev.iadev.domain.stack;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.assertj.core.api.Assertions.assertThat;

@DisplayName("StackMapping (Java-only since EPIC-0048 / v4.0.0)")
class StackMappingTest {

    @Nested
    @DisplayName("LANGUAGE_COMMANDS")
    class LanguageCommandsTests {

        @Test
        @DisplayName("contains exactly 2 Java entries")
        void languageCommands_size_two() {
            assertThat(StackMapping.LANGUAGE_COMMANDS).hasSize(2);
        }

        @ParameterizedTest
        @ValueSource(strings = {"java-maven", "java-gradle"})
        @DisplayName("contains key: {0}")
        void languageCommands_containsKey(String key) {
            assertThat(StackMapping.LANGUAGE_COMMANDS).containsKey(key);
        }

        @Test
        @DisplayName("java-maven has correct commands")
        void languageCommands_javaMaven_correctCommands() {
            var cmds = StackMapping.LANGUAGE_COMMANDS.get("java-maven");
            assertThat(cmds.compileCmd()).isEqualTo("./mvnw compile -q");
            assertThat(cmds.buildCmd())
                    .isEqualTo("./mvnw package -DskipTests");
            assertThat(cmds.testCmd()).isEqualTo("./mvnw verify");
            assertThat(cmds.coverageCmd())
                    .isEqualTo("./mvnw verify jacoco:report");
            assertThat(cmds.fileExtension()).isEqualTo(".java");
            assertThat(cmds.buildFile()).isEqualTo("pom.xml");
            assertThat(cmds.packageManager()).isEqualTo("maven");
        }

        @Test
        @DisplayName("java-gradle has correct commands")
        void languageCommands_javaGradle_correctCommands() {
            var cmds = StackMapping.LANGUAGE_COMMANDS.get("java-gradle");
            assertThat(cmds.compileCmd())
                    .isEqualTo("./gradlew compileJava -q");
            assertThat(cmds.fileExtension()).isEqualTo(".java");
            assertThat(cmds.buildFile()).isEqualTo("build.gradle");
            assertThat(cmds.packageManager()).isEqualTo("gradle");
        }

        @ParameterizedTest
        @ValueSource(strings = {
                "kotlin-gradle", "typescript-npm", "python-pip",
                "go-go", "rust-cargo", "csharp-dotnet"})
        @DisplayName("non-Java key {0} is absent")
        void languageCommands_nonJavaKey_absent(String key) {
            assertThat(StackMapping.LANGUAGE_COMMANDS)
                    .doesNotContainKey(key);
        }
    }

    @Nested
    @DisplayName("FRAMEWORK_PORTS")
    class FrameworkPortsTests {

        @Test
        @DisplayName("contains exactly 2 Java entries")
        void frameworkPorts_size_two() {
            assertThat(StackMapping.FRAMEWORK_PORTS).hasSize(2);
        }

        @Test
        @DisplayName("quarkus port is 8080")
        void frameworkPorts_quarkus() {
            assertThat(StackMapping.FRAMEWORK_PORTS.get("quarkus"))
                    .isEqualTo(8080);
        }

        @Test
        @DisplayName("spring-boot port is 8080")
        void frameworkPorts_springBoot() {
            assertThat(StackMapping.FRAMEWORK_PORTS.get("spring-boot"))
                    .isEqualTo(8080);
        }

        @ParameterizedTest
        @ValueSource(strings = {
                "nestjs", "express", "fastapi", "django",
                "gin", "ktor", "axum", "actix-web", "aspnet"})
        @DisplayName("non-Java framework {0} is absent")
        void frameworkPorts_nonJava_absent(String fw) {
            assertThat(StackMapping.FRAMEWORK_PORTS)
                    .doesNotContainKey(fw);
        }
    }

    @Nested
    @DisplayName("FRAMEWORK_HEALTH_PATHS")
    class FrameworkHealthPathsTests {

        @Test
        @DisplayName("contains exactly 2 Java entries")
        void frameworkHealthPaths_size_two() {
            assertThat(StackMapping.FRAMEWORK_HEALTH_PATHS).hasSize(2);
        }

        @Test
        @DisplayName("quarkus has /q/health")
        void frameworkHealthPaths_whenCalled_quarkus() {
            assertThat(StackMapping.FRAMEWORK_HEALTH_PATHS.get("quarkus"))
                    .isEqualTo("/q/health");
        }

        @Test
        @DisplayName("spring-boot has /actuator/health")
        void frameworkHealthPaths_whenCalled_springBoot() {
            assertThat(StackMapping.FRAMEWORK_HEALTH_PATHS.get("spring-boot"))
                    .isEqualTo("/actuator/health");
        }
    }

    @Nested
    @DisplayName("FRAMEWORK_LANGUAGE_RULES")
    class FrameworkLanguageRulesTests {

        @Test
        @DisplayName("contains 2 Java entries")
        void frameworkLanguageRules_size_two() {
            assertThat(StackMapping.FRAMEWORK_LANGUAGE_RULES).hasSize(2);
        }

        @Test
        @DisplayName("quarkus accepts java")
        void frameworkLanguageRules_whenCalled_quarkus() {
            assertThat(StackMapping.FRAMEWORK_LANGUAGE_RULES.get("quarkus"))
                    .containsExactly("java");
        }

        @Test
        @DisplayName("spring-boot accepts java")
        void frameworkLanguageRules_whenCalled_springBoot() {
            assertThat(
                    StackMapping.FRAMEWORK_LANGUAGE_RULES.get("spring-boot"))
                    .containsExactly("java");
        }

        @ParameterizedTest
        @ValueSource(strings = {
                "nestjs", "fastapi", "gin", "ktor", "axum",
                "aspnet", "django", "express", "fastify",
                "commander", "flask", "stdlib", "fiber", "actix-web"})
        @DisplayName("non-Java framework {0} is absent")
        void frameworkLanguageRules_nonJava_absent(String fw) {
            assertThat(StackMapping.FRAMEWORK_LANGUAGE_RULES)
                    .doesNotContainKey(fw);
        }
    }

    @Nested
    @DisplayName("Constants")
    class ConstantsTests {

        @Test
        @DisplayName("NATIVE_SUPPORTED_FRAMEWORKS has quarkus and spring-boot")
        void nativeSupportedFrameworks_whenCalled_twoEntries() {
            assertThat(StackMapping.NATIVE_SUPPORTED_FRAMEWORKS)
                    .containsExactly("quarkus", "spring-boot");
        }

        @Test
        @DisplayName("VALID_INTERFACE_TYPES has 9 entries")
        void validInterfaceTypes_whenCalled_nine() {
            assertThat(StackMapping.VALID_INTERFACE_TYPES).hasSize(9);
            assertThat(StackMapping.VALID_INTERFACE_TYPES)
                    .contains("rest", "grpc", "graphql", "websocket",
                            "tcp-custom", "cli", "event-consumer",
                            "event-producer", "scheduled");
        }

        @Test
        @DisplayName("VALID_ARCHITECTURE_STYLES has 10 entries")
        void validArchitectureStyles_whenCalled_ten() {
            assertThat(StackMapping.VALID_ARCHITECTURE_STYLES)
                    .hasSize(10);
        }

        @Test
        @DisplayName("DEFAULT_PORT_FALLBACK is 8080")
        void map_whenCalled_defaultPortFallback() {
            assertThat(StackMapping.DEFAULT_PORT_FALLBACK).isEqualTo(8080);
        }

        @Test
        @DisplayName("DEFAULT_HEALTH_PATH is /health")
        void map_whenCalled_defaultHealthPath() {
            assertThat(StackMapping.DEFAULT_HEALTH_PATH)
                    .isEqualTo("/health");
        }

        @Test
        @DisplayName("DEFAULT_DOCKER_IMAGE is alpine:latest")
        void map_whenCalled_defaultDockerImage() {
            assertThat(StackMapping.DEFAULT_DOCKER_IMAGE)
                    .isEqualTo("alpine:latest");
        }

        @Test
        @DisplayName("DOCKER_BASE_IMAGES has 1 Java entry")
        void dockerBaseImages_whenCalled_one() {
            assertThat(StackMapping.DOCKER_BASE_IMAGES).hasSize(1);
            assertThat(StackMapping.DOCKER_BASE_IMAGES)
                    .containsKey("java");
        }

        @Test
        @DisplayName("INTERFACE_SPEC_PROTOCOL_MAP has 7 entries")
        void interfaceSpecProtocolMap_whenCalled_seven() {
            assertThat(StackMapping.INTERFACE_SPEC_PROTOCOL_MAP).hasSize(7);
        }

        @Test
        @DisplayName("HOOK_TEMPLATE_MAP has 2 Java entries")
        void hookTemplateMap_whenCalled_two() {
            assertThat(StackMapping.HOOK_TEMPLATE_MAP).hasSize(2);
        }

        @Test
        @DisplayName("SETTINGS_LANG_MAP has 2 Java entries")
        void settingsLangMap_whenCalled_two() {
            assertThat(StackMapping.SETTINGS_LANG_MAP).hasSize(2);
        }
    }

    @Nested
    @DisplayName("Utility Methods")
    class UtilityMethodsTests {

        @Test
        @DisplayName("getHookTemplateKey returns correct key")
        void getHookTemplateKey_javaGradle_correct() {
            assertThat(StackMapping.getHookTemplateKey("java", "gradle"))
                    .isEqualTo("java-gradle");
        }

        @Test
        @DisplayName("getHookTemplateKey returns empty for python")
        void getHookTemplateKey_pythonPip_empty() {
            assertThat(StackMapping.getHookTemplateKey("python", "pip"))
                    .isEmpty();
        }

        @Test
        @DisplayName("getHookTemplateKey returns empty for unknown")
        void getHookTemplateKey_unknown_empty() {
            assertThat(StackMapping.getHookTemplateKey("unknown", "x"))
                    .isEmpty();
        }

        @Test
        @DisplayName("getSettingsLangKey returns correct key")
        void getSettingsLangKey_javaMaven_correct() {
            assertThat(StackMapping.getSettingsLangKey("java", "maven"))
                    .isEqualTo("java-maven");
        }

        @Test
        @DisplayName("getSettingsLangKey returns empty for unknown")
        void getSettingsLangKey_unknown_empty() {
            assertThat(StackMapping.getSettingsLangKey("unknown", "x"))
                    .isEmpty();
        }
    }
}
